components/cam/src/dynamics/se/fv_physics_coupling_mod.F90

```diff
@@ -0,0 +1,284 @@
+module fv_physics_coupling_mod
+  use element_mod,    only: element_t
+  use shr_kind_mod,   only: r8=>shr_kind_r8
+  use kinds,          only: real_kind, int_kind
+  use constituents,   only: pcnst, cnst_name
+  use dimensions_mod, only: np, npsq, nelemd, nlev
+  use dyn_grid,       only: fv_nphys
+  use ppgrid,         only: pcols, pver, pverp
+  
+  private
+
+  ! These method encapsulate the coupling method for the fv phys grid,
+  ! they are used by d_p_coupling() and p_d_coupling() in dp_coupling.F90 
+  public :: fv_phys_to_dyn
+  public :: fv_phys_to_dyn_topo
+  public :: dyn_to_fv_phys
+
+contains
+  !=================================================================================================
+  !=================================================================================================
+  subroutine fv_phys_to_dyn(elem,T_tmp,uv_tmp,q_tmp)
+    ! Purpose: Copy physics state to dynamics grid
+    use control_mod,    only: ftype
+    use dyn_comp,       only: TimeLevel
+    use derivative_mod, only: subcell_integration
+    implicit none
+    !---------------------------------------------------------------------------
+    ! interface arguments
+    type(element_t),      intent(inout) :: elem(:)          ! dynamics element structure
+    real(kind=real_kind), intent(inout) :: T_tmp (:,:,:)    ! temp array to hold T
+    real(kind=real_kind), intent(inout) :: uv_tmp(:,:,:,:)  ! temp array to hold u and v
+    real(kind=real_kind), intent(inout) :: q_tmp (:,:,:,:)  ! temp to hold advected constituents
+    ! local variables
+    integer(kind=int_kind) :: ie, m, i, j, icol, ilyr       ! loop iterators
+    integer(kind=int_kind) :: ii, jj, gi, gj                ! GLL loop iterator and indices for pg2
+    integer                :: tl_f
+    real(kind=real_kind), dimension(fv_nphys*fv_nphys,pver,pcnst) :: qo_phys ! reconstructed initial physics state 
+    real(r8), dimension(np,np)             :: dp_gll
+    real(r8), dimension(fv_nphys,fv_nphys) :: inv_dp_fvm
+    !---------------------------------------------------------------------------
+    ! Copy tendencies on the physics grid over to the dynamics grid (GLL)
+    !---------------------------------------------------------------------------
+    tl_f = TimeLevel%n0
+    do ie = 1,nelemd
+      !-------------------------------------------------------------------------
+      ! Recalculate state that was previously sent to physics 
+      !-------------------------------------------------------------------------
+      if (ftype==2.or.ftype==4) then
+        do ilyr = 1,pver
+          dp_gll = elem(ie)%state%dp3d(:,:,ilyr,tl_f)
+          inv_dp_fvm = 1.0 / subcell_integration(dp_gll,np,fv_nphys,elem(ie)%metdet(:,:))
+          do m = 1,pcnst
+            qo_phys(:,ilyr,m)  = RESHAPE( subcell_integration(              &
+                                  elem(ie)%state%Q(:,:,ilyr,m)*dp_gll,      &
```

Comment:
Q*dp is stored in state%Qdp variable which I think is not overwritten by physics? so, maybe multiplication `elem(ie)%state%Q(:,:,ilyr,m)*dp_gll` is not needed?

Reply:
I have to calculate dp_gll either way for temperature, but I could change the Q part to Qdp. My only concern is whether Qdp is derived from Q? Or is it the other way around? Or should I handle both here?

Reply:
After thinking about this more, I think I'd prefer to keep it how it is now to be sure that the normalization by dp is correct.

Reply:
It is an extra operation that may not be needed?

Reply:
maybe, but this is the safest way to go unless we can be 100% sure that Qdp is exactly equal to the same Q*dp that's calculated here, and I'm not sure how best to check that.